Karaf: system:shutdown - Callback is missing

Hi,
shutting down OH2 via Karaf result in some errors:
In my opinion this is because i linked the Items manuell in my default.items file.
So that's really no error at all. Wish to get rid of this "error-messages" in log file because its irritatingly confusing.

```
18:20:33.855 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occured while disposing handler of thing 'max:thermostat:MKF0071697:MKF0048314': java.lang.IllegalStateException: Could not update status, because callback is missing
java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Could not update status, because callback is missing
    at java.util.concurrent.FutureTask.report(FutureTask.java:122)[:1.8.0_65]
    at java.util.concurrent.FutureTask.get(FutureTask.java:206)[:1.8.0_65]
    at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:179)
    at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:72)
    at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:56)
    at org.eclipse.smarthome.core.thing.internal.ThingManager.disposeHandler(ThingManager.java:738)
    at org.eclipse.smarthome.core.thing.internal.ThingManager.handlerRemoved(ThingManager.java:340)
    at org.eclipse.smarthome.core.thing.internal.ThingManager$ThingHandlerTracker.removedService(ThingManager.java:122)
    at org.eclipse.smarthome.core.thing.internal.ThingManager$ThingHandlerTracker.removedService(ThingManager.java:1)
    at org.osgi.util.tracker.ServiceTracker$Tracked.customizerRemoved(ServiceTracker.java:967)[org.osgi.core-6.0.0.jar:]
    at org.osgi.util.tracker.ServiceTracker$Tracked.customizerRemoved(ServiceTracker.java:870)[org.osgi.core-6.0.0.jar:]
    at org.osgi.util.tracker.AbstractTracked.untrack(AbstractTracked.java:341)[org.osgi.core-6.0.0.jar:]
    at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:909)[org.osgi.core-6.0.0.jar:]
    at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:109)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:914)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:862)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:801)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.unregister(ServiceRegistrationImpl.java:222)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory.unregisterHandler(BaseThingHandlerFactory.java:105)
    at org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory.deactivate(BaseThingHandlerFactory.java:74)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)[:1.8.0_65]
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)[:1.8.0_65]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)[:1.8.0_65]
    at java.lang.reflect.Method.invoke(Method.java:497)[:1.8.0_65]
    at org.apache.felix.scr.impl.helper.BaseMethod.invokeMethod(BaseMethod.java:222)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.helper.BaseMethod.access$500(BaseMethod.java:37)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.helper.BaseMethod$Resolved.invoke(BaseMethod.java:615)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.helper.BaseMethod.invoke(BaseMethod.java:499)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.helper.ActivateMethod.invoke(ActivateMethod.java:295)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.manager.SingleComponentManager.disposeImplementationObject(SingleComponentManager.java:342)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.manager.SingleComponentManager.deleteComponent(SingleComponentManager.java:157)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.manager.AbstractComponentManager.doDeactivate(AbstractComponentManager.java:783)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.manager.AbstractComponentManager.deactivateInternal(AbstractComponentManager.java:757)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.manager.AbstractComponentManager.dispose(AbstractComponentManager.java:550)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.config.ConfigurableComponentHolder.disposeComponents(ConfigurableComponentHolder.java:725)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.BundleComponentActivator.dispose(BundleComponentActivator.java:530)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.Activator.disposeComponents(Activator.java:414)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.Activator.access$300(Activator.java:53)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.scr.impl.Activator$ScrExtension.destroy(Activator.java:273)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.utils.extender.AbstractExtender$2.run(AbstractExtender.java:290)[29:org.apache.felix.scr:2.0.2]
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_65]
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_65]
    at org.apache.felix.utils.extender.AbstractExtender.destroyExtension(AbstractExtender.java:312)[29:org.apache.felix.scr:2.0.2]
    at org.apache.felix.utils.extender.AbstractExtender.bundleChanged(AbstractExtender.java:186)[29:org.apache.felix.scr:2.0.2]
    at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:902)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:148)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:165)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:75)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:67)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:102)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.container.Module.publishEvent(Module.java:466)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.container.Module.doStop(Module.java:624)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.container.Module.stop(Module.java:488)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.decStartLevel(ModuleContainer.java:1623)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1542)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.container.SystemModule.stopWorker(SystemModule.java:248)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.internal.framework.EquinoxBundle$SystemBundle$EquinoxSystemModule.stopWorker(EquinoxBundle.java:145)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.container.Module.doStop(Module.java:626)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.container.Module.stop(Module.java:488)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.container.SystemModule.stop(SystemModule.java:186)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at org.eclipse.osgi.internal.framework.EquinoxBundle$SystemBundle$EquinoxSystemModule$1.run(EquinoxBundle.java:160)[org.eclipse.osgi-3.10.2.v20150203-1939.jar:]
    at java.lang.Thread.run(Thread.java:745)[:1.8.0_65]
Caused by: java.lang.IllegalStateException: Could not update status, because callback is missing
    at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:307)[95:org.eclipse.smarthome.core.thing:0.8.0.201604051458]
    at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:334)[95:org.eclipse.smarthome.core.thing:0.8.0.201604051458]
    at org.openhab.binding.max.internal.handler.MaxDevicesHandler.dispose(MaxDevicesHandler.java:142)[153:org.openhab.binding.max:2.0.0.201603310102]
    at org.eclipse.smarthome.core.thing.internal.ThingManager$10.call(ThingManager.java:741)[95:org.eclipse.smarthome.core.thing:0.8.0.201604051458]
    at org.eclipse.smarthome.core.thing.internal.ThingManager$10.call(ThingManager.java:1)[95:org.eclipse.smarthome.core.thing:0.8.0.201604051458]
    at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:170)[89:org.eclipse.smarthome.core:0.8.0.201604051458]
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_65]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_65]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_65]
    ... 1 more
```

Andi

--- a/bundles/core/org.eclipse.smarthome.core.thing.test/src/test/groovy/org/eclipse/smarthome/core/thing/internal/ThingManagerOSGiTest.groovy
+++ b/bundles/core/org.eclipse.smarthome.core.thing.test/src/test/groovy/org/eclipse/smarthome/core/thing/internal/ThingManagerOSGiTest.groovy
@@ -365,30 +365,32 @@ class ThingManagerOSGiTest extends OSGiTest {
     void 'ThingManager handles bridge-thing handler life cycle correctly'() {
         initCalledCounter = 0
         disposedCalledCounter = 0
-        ThingHandlerCallback callback
 
         def bridge = BridgeBuilder.create(new ThingTypeUID("binding:test"), new ThingUID("binding:test:someBridgeUID-1")).build()
         def bridgeInitCalled = false
         def bridgeInitCalledOrder = 0
         def bridgeDisposedCalled = false
         def bridgeDisposedCalledOrder = 0
+        def ThingHandlerCallback bridgeCallback
+        def bridgeCallbackWasNull = false
         def bridgeHandler = [
-            setCallback: { callbackArg -> callback = callbackArg },
+            setCallback: { callbackArg -> bridgeCallback = callbackArg },
             initialize: {
                 bridgeInitCalled = true
                 bridgeInitCalledOrder = ++initCalledCounter
-                callback.statusUpdated(bridge, ThingStatusInfoBuilder.create(ThingStatus.ONLINE, ThingStatusDetail.NONE).build())
+                bridgeCallback.statusUpdated(bridge, ThingStatusInfoBuilder.create(ThingStatus.ONLINE, ThingStatusDetail.NONE).build())
             },
             dispose: {
                 bridgeDisposedCalled = true
                 bridgeDisposedCalledOrder = ++disposedCalledCounter
+                bridgeCallbackWasNull = bridgeCallback == null
             },
             thingDisposed: { thing -> },
             childHandlerInitialized: { thingHandler, thing -> },
             childHandlerDisposed: { thingHandler, thing -> },
             getThing: { bridge },
             handleRemoval: {
-                callback.statusUpdated(bridge, ThingStatusInfoBuilder.create(ThingStatus.REMOVED).build())
+                bridgeCallback.statusUpdated(bridge, ThingStatusInfoBuilder.create(ThingStatus.REMOVED).build())
             }
         ] as BridgeHandler
 
@@ -397,21 +399,24 @@ class ThingManagerOSGiTest extends OSGiTest {
         def thingInitCalledOrder = 0
         def thingDisposedCalled = false
         def thingDisposedCalledOrder = 0
+        def ThingHandlerCallback thingCallback
+        def thingCallbackWasNull = false
         def thingHandler = [
-            setCallback: {},
+            setCallback: { callbackArg -> thingCallback = callbackArg },
             initialize: {
                 thingInitCalled = true
                 thingInitCalledOrder = ++initCalledCounter
-                callback.statusUpdated(thing, ThingStatusInfoBuilder.create(ThingStatus.ONLINE, ThingStatusDetail.NONE).build())
+                bridgeCallback.statusUpdated(thing, ThingStatusInfoBuilder.create(ThingStatus.ONLINE, ThingStatusDetail.NONE).build())
             },
             dispose: {
                 thingDisposedCalled = true
                 thingDisposedCalledOrder = ++disposedCalledCounter
+                thingCallbackWasNull = thingCallback == null
             },
             bridgeStatusChanged: { },
             getThing: { thing },
             handleRemoval: {
-                callback.statusUpdated(thing, ThingStatusInfoBuilder.create(ThingStatus.REMOVED).build())
+                bridgeCallback.statusUpdated(thing, ThingStatusInfoBuilder.create(ThingStatus.REMOVED).build())
             }
         ] as ThingHandler
 
@@ -465,6 +470,8 @@ class ThingManagerOSGiTest extends OSGiTest {
         waitForAssert ({ assertThat bridge.getStatusInfo(), is(statusInfo) })
 
         // remove thing - provokes thing disposal
+        bridgeCallbackWasNull = false
+        thingCallbackWasNull = false
         thingRegistry.remove(thing.getUID())
         waitForAssert ({ assertThat thingDisposedCalled, is(true) })
         waitForAssert ({ assertThat thingDisposedCalledOrder, is(1) })
@@ -472,6 +479,7 @@ class ThingManagerOSGiTest extends OSGiTest {
         waitForAssert ({ assertThat bridge.getStatusInfo(), is(statusInfo) })
         statusInfo = ThingStatusInfoBuilder.create(ThingStatus.UNINITIALIZED, ThingStatusDetail.HANDLER_MISSING_ERROR).build()
         waitForAssert ({ assertThat thing.getStatusInfo(), is(statusInfo) })
+        assertThat thingCallbackWasNull, is(false)
 
         // add thing again - provokes thing initialization
         thingRegistry.add(thing)
@@ -482,6 +490,8 @@ class ThingManagerOSGiTest extends OSGiTest {
         waitForAssert ({ assertThat thing.getStatusInfo(), is(statusInfo) })
 
         // remove bridge - provokes thing & bridge disposal
+        bridgeCallbackWasNull = false
+        thingCallbackWasNull = false
         thingRegistry.remove(bridge.getUID())
         waitForAssert ({ assertThat thingDisposedCalled, is(true) })
         waitForAssert ({ assertThat thingDisposedCalledOrder, is(2) })
@@ -492,6 +502,8 @@ class ThingManagerOSGiTest extends OSGiTest {
         statusInfo = ThingStatusInfoBuilder.create(ThingStatus.UNINITIALIZED, ThingStatusDetail.HANDLER_MISSING_ERROR).build()
         waitForAssert ({ assertThat thing.getStatusInfo(), is(statusInfo) })
         waitForAssert ({ assertThat bridge.getStatusInfo(), is(statusInfo) })
+        assertThat bridgeCallbackWasNull, is(false)
+        assertThat thingCallbackWasNull, is(false)
 
         // add bridge again
         thingRegistry.add(bridge)
@@ -506,6 +518,8 @@ class ThingManagerOSGiTest extends OSGiTest {
         waitForAssert ({ assertThat bridge.getStatusInfo(), is(statusInfo) })
 
         // unregister factory
+        bridgeCallbackWasNull = false
+        thingCallbackWasNull = false
         unregisterService(thingHandlerFactory)
         waitForAssert ({ assertThat thingDisposedCalled, is(true) })
         waitForAssert ({ assertThat thingDisposedCalledOrder, is(4) })
@@ -514,6 +528,8 @@ class ThingManagerOSGiTest extends OSGiTest {
         statusInfo = ThingStatusInfoBuilder.create(ThingStatus.UNINITIALIZED, ThingStatusDetail.HANDLER_MISSING_ERROR).build()
         waitForAssert ({ assertThat thing.getStatusInfo(), is(statusInfo) })
         waitForAssert ({ assertThat bridge.getStatusInfo(), is(statusInfo) })
+        assertThat bridgeCallbackWasNull, is(false)
+        assertThat thingCallbackWasNull, is(false)
     }
 
     @Test

--- a/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java
+++ b/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java
@@ -283,8 +283,7 @@ public class ThingManager extends AbstractItemEventSubscriber implements ThingTr
                 final ThingHandlerFactory oldThingHandlerFactory = findThingHandlerFactory(thing.getThingTypeUID());
                 if (oldThingHandlerFactory != null) {
                     ThingHandler thingHandler = thing.getHandler();
-                    unregisterHandler(thing, oldThingHandlerFactory);
-                    disposeHandler(thing, thingHandler);
+                    unregisterAndDisposeHandler(oldThingHandlerFactory, thing, thingHandler);
                     waitUntilHandlerUnregistered(thing, 60 * 1000);
                 } else {
                     logger.debug("No ThingHandlerFactory available that can handle {}", thing.getThingTypeUID());
@@ -476,8 +475,7 @@ public class ThingManager extends AbstractItemEventSubscriber implements ThingTr
         if (thingHandler != null) {
             final ThingHandlerFactory thingHandlerFactory = findThingHandlerFactory(thing.getThingTypeUID());
             if (thingHandlerFactory != null) {
-                unregisterHandler(thing, thingHandlerFactory);
-                disposeHandler(thing, thingHandler);
+                unregisterAndDisposeHandler(thingHandlerFactory, thing, thingHandler);
                 if (thingTrackerEvent == ThingTrackerEvent.THING_REMOVED) {
                     SafeMethodCaller.call(new SafeMethodCaller.Action<Void>() {
                         @Override
@@ -754,12 +752,7 @@ public class ThingManager extends AbstractItemEventSubscriber implements ThingTr
     private void unregisterHandler(Thing thing, ThingHandlerFactory thingHandlerFactory) {
         synchronized (thing) {
             if (isHandlerRegistered(thing)) {
-                if (!isBridge(thing)) {
-                    doUnregisterHandler(thing, thingHandlerFactory);
-                } else {
-                    unregisterChildHandlers((Bridge) thing, thingHandlerFactory);
-                    doUnregisterHandler(thing, thingHandlerFactory);
-                }
+                doUnregisterHandler(thing, thingHandlerFactory);
             }
         }
     }
@@ -802,6 +795,8 @@ public class ThingManager extends AbstractItemEventSubscriber implements ThingTr
             SafeMethodCaller.call(new SafeMethodCaller.ActionWithException<Void>() {
                 @Override
                 public Void call() throws Exception {
+                    setThingStatus(thingHandler.getThing(),
+                            buildStatusInfo(ThingStatus.UNINITIALIZED, ThingStatusDetail.NONE));
                     thingHandler.dispose();
                     return null;
                 }
@@ -815,16 +810,24 @@ public class ThingManager extends AbstractItemEventSubscriber implements ThingTr
         }
     }
 
-    private void unregisterChildHandlers(Bridge bridge, ThingHandlerFactory thingHandlerFactory) {
+    private void unregisterAndDisposeChildHandlers(Bridge bridge, ThingHandlerFactory thingHandlerFactory) {
         addThingsToBridge(bridge);
         for (Thing child : bridge.getThings()) {
             ThingHandler handler = child.getHandler();
             if (handler != null) {
                 logger.debug("Unregister and dispose child '{}' of bridge '{}'.", child.getUID(), bridge.getUID());
-                unregisterHandler(child, thingHandlerFactory);
-                disposeHandler(child, handler);
+                unregisterAndDisposeHandler(thingHandlerFactory, child, handler);
             }
         }
+    }
+
+    private void unregisterAndDisposeHandler(ThingHandlerFactory thingHandlerFactory, Thing thing,
+            ThingHandler handler) {
+        if (isBridge(thing)) {
+            unregisterAndDisposeChildHandlers((Bridge) thing, thingHandlerFactory);
+        }
+        disposeHandler(thing, handler);
+        unregisterHandler(thing, thingHandlerFactory);
     }
 
     private void addThingsToBridge(Bridge bridge) {
@@ -1021,8 +1024,7 @@ public class ThingManager extends AbstractItemEventSubscriber implements ThingTr
         for (ThingHandler thingHandler : handlers) {
             Thing thing = thingHandler.getThing();
             if (thing != null && isHandlerRegistered(thing)) {
-                unregisterHandler(thing, thingHandlerFactory);
-                disposeHandler(thing, thingHandler);
+                unregisterAndDisposeHandler(thingHandlerFactory, thing, thingHandler);
             }
         }
         thingHandlersByFactory.removeAll(thingHandlerFactory);

